File: src/test/java/pl/training/workshop/m6/s19_visitor/S19SolutionTest.java
package pl.training.workshop.m6.s19_visitor;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.List;

import org.junit.jupiter.api.Test;

import pl.training.workshop.m6.s19_visitor.step2.OrderItem;
import pl.training.workshop.m6.s19_visitor.step2.OrderItemVisitor;
import pl.training.workshop.m6.s19_visitor.step2.SnackItem;
import pl.training.workshop.m6.s19_visitor.step2.TicketItem;
import pl.training.workshop.m6.s19_visitor.step2.VoucherItem;
import pl.training.workshop.shared.Money;

/** Macierz zmian: nowa operacja jest tania zarówno jako Visitor, jak i jako switch po sealed. */
final class S19SolutionTest {
    @Test
    void newOperationIsANewVisitor() {
        OrderItemVisitor<String> vatGroup = new OrderItemVisitor<>() {
            @Override
            public String visitTicket(TicketItem ticket) {
                return "B";
            }

            @Override
            public String visitSnack(SnackItem snack) {
                return "A";
            }

            @Override
            public String visitVoucher(VoucherItem voucher) {
                return "-";
            }
        };
        List<OrderItem> items = List.of(new TicketItem("Diuna", "IMAX", Money.of("40.00")),
                new SnackItem("Cola", Money.of("9.00")), new VoucherItem("KINO20", Money.of("20.00")));
        assertEquals(List.of("B", "A", "-"), items.stream().map(item -> item.accept(vatGroup)).toList());
    }

    @Test
    void newOperationIsANewSwitchInJava25() {
        var items = List.<pl.training.workshop.m6.s19_visitor.step3.OrderItem>of(
                new pl.training.workshop.m6.s19_visitor.step3.TicketItem("Diuna", "IMAX", Money.of("40.00")),
                new pl.training.workshop.m6.s19_visitor.step3.VoucherItem("KINO20", Money.of("20.00")));
        assertEquals(List.of("B", "-"), items.stream().map(item -> switch (item) {
            case pl.training.workshop.m6.s19_visitor.step3.TicketItem ticket -> "B";
            case pl.training.workshop.m6.s19_visitor.step3.SnackItem snack -> "A";
            case pl.training.workshop.m6.s19_visitor.step3.VoucherItem voucher -> "-";
        }).toList());
    }
}

File: csharp/tests/Training.Workshop.Tests/M6/S19Visitor/S19SolutionTest.cs
using System.Diagnostics;
using Training.Workshop.M6.S19Visitor.Step2;
using Training.Workshop.Shared;

namespace Training.Workshop.Tests.M6.S19Visitor;

/// <summary>Macierz zmian: nowa operacja jest tania zarówno jako Visitor, jak i jako switch po typach.</summary>
public sealed class S19SolutionTest
{
    [Fact]
    public void NewOperationIsANewVisitor()
    {
        var vatGroup = new VatGroupVisitor();
        IReadOnlyList<IOrderItem> items = [new TicketItem("Diuna", "IMAX", Money.Of("40.00")),
            new SnackItem("Cola", Money.Of("9.00")), new VoucherItem("KINO20", Money.Of("20.00"))];
        Assert.Equal(["B", "A", "-"], items.Select(item => item.Accept(vatGroup)).ToList());
    }

    /// <summary>Odpowiednik newOperationIsANewSwitchInJava25 - w C# switch expression z ramieniem "_".</summary>
    [Fact]
    public void NewOperationIsANewSwitchExpression()
    {
        IReadOnlyList<Training.Workshop.M6.S19Visitor.Step3.IOrderItem> items = [
            new Training.Workshop.M6.S19Visitor.Step3.TicketItem("Diuna", "IMAX", Money.Of("40.00")),
            new Training.Workshop.M6.S19Visitor.Step3.VoucherItem("KINO20", Money.Of("20.00"))];
        Assert.Equal(["B", "-"], items.Select(item => item switch
        {
            Training.Workshop.M6.S19Visitor.Step3.TicketItem => "B",
            Training.Workshop.M6.S19Visitor.Step3.SnackItem => "A",
            Training.Workshop.M6.S19Visitor.Step3.VoucherItem => "-",
            _ => throw new UnreachableException(),
        }).ToList());
    }

    private sealed class VatGroupVisitor : IOrderItemVisitor<string>
    {
        public string VisitTicket(TicketItem ticket) => "B";

        public string VisitSnack(SnackItem snack) => "A";

        public string VisitVoucher(VoucherItem voucher) => "-";
    }
}
